torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+        length_tensor = torch.LongTensor([len(data)]).to(dist_device)
+        data_send_tensor = torch.ByteTensor(data).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        dist.broadcast(data_send_tensor, src=src_rank, group=group, async_op=False)
+    else:
+        # Fetch from the source
+        length_tensor = torch.LongTensor([0]).to(dist_device)
+        dist.broadcast(length_tensor, src=src_rank, group=group, async_op=False)
+        data_recv_tensor = torch.empty([int(length_tensor.item())], dtype=torch.uint8, device=dist_device)
+        dist.broadcast(data_recv_tensor, src=src_rank, group=group, async_op=False)
+        buffer = io.BytesIO(data_recv_tensor.cpu().numpy())
+        obj = torch.load(buffer, map_location=dist_device)
+    return obj
+
+
+# Credits:  classy_vision/generic/distributed_util.py
+def _recursive_copy_to_device(value: Any, non_blocking: bool, device: torch.device) -> Any:
+    """
+    Recursively searches lists, tuples, dicts and copies tensors to device if
+    possible. Non-tensor values are passed as-is in the result.
+
+    NOTE:  These are all copies, so if there are two objects that reference
+    the same object, then after this call, there will be two different objects
+    referenced on the device.
+    """
+
+    if isinstance(value, torch.Tensor):
+        return value.to(device, non_blocking=non_blocking)
+
+    if isinstance(value, (list, tuple)):
+        values = [_recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for val in value]
+        return values if isinstance(value, list) else tuple(values)
+
+    if isinstance(value, container_abcs.Mapping):
+        return {
+            key: _recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for key, val in value.items()
+        }
+
+    return value
+
+
+def _get_global_rank(group: Any, rank: int) -> int:
+    if group is dist.group.WORLD:
+        return rank
+    else:
+        global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+    return global_rank
+
+
+class ZeROptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ShardedOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters
+    at each rank. Each parameter belongs to a single rank and
+    is not divided among rank.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
+    Args:
+        params (list of tensors):
+            parameters to be optimized
+    Keyword Args:
+        optim (torch.nn.Optimizer):
+            optimizer to shard
+        group (group):
```

Comment:
Since this one and the the one below (`broadcasat_buffer_size`) are optional, shall we make it clear in the doc? An example is listed below:

https://github.com/pytorch/pytorch/blob/b5662ba0f0cb6d99805a44dd5b6bcda128468c83/torch/distributed/rpc/api.py#L428